### ISSUE TITLE
CollectionViewDiffCalculator crashes every time

### ISSUE THREAD

**Issue by bondareb-evgenii** [jflinter/Dwifft#21]
Fix taken from here: http://stackoverflow.com/questions/25265183/collectionview-performbatchupdates-crash

**if oldRows.isEmpty {
                    collectionView?.reloadData()
                }** else {
                    collectionView?.performBatchUpdates({ () -> Void in
                        self.collectionView?.insertItemsAtIndexPaths(insertionIndexPaths)
                        self.collectionView?.deleteItemsAtIndexPaths(deletionIndexPaths)
                    }, completion: nil)
                }

**Comment by jflinter** [comment]
Can you provide a bit more context here? What version of iOS are you observing this crash on? Could you link to a small project that reproduces this? The fact that others have used this code safely makes me wonder about the actual root cause here.

**Comment by elliottwilliams** [comment]
I think this is the same general problem I mentioned in #23: UIKit expects the data source's counts to change within `performBatchUpdates`. I saw the same `NSInternalInconsistencyException` errors I documented in #23.

I need to pull the latest master (and migrate to Swift 3.0), but I should be able to fix this by changing `CollectionViewDiffCalculator` to use a `set` block instead of `didSet`. `set` will calculate the diff and, if the diff introduces changes, call `performBatchUpdates`, update the data source, and request the add/remove animations.
